File: quick_program/src/states/Bond.rs
use borsh::{BorshDeserialize, BorshSerialize};


#[derive(BorshSerialize, BorshDeserialize, Debug, Clone)]
pub struct Bond{
    pub name: Vec<u8>,
    pub amount: u64,
    pub rate: u16,
    pub used: u16,
    pub sold: u16
}

impl Bond {
    pub fn new(name: Vec<u8>, amount: u64, rate: u16) -> Self {
        Bond{name, amount, rate, used: 0, sold: 0}
    }
}

File: helpers/src/schemas/bond.ts
import {TString} from "./string";
import {TDeserializable} from "./vector";
import BigInteger from "big-integer";
import {fromLittleEndian} from "../big-byte";
import BN from "bn.js";


export interface IBond {
    name: TString,
    amount: BigInteger,
    rate: number,
    used: number,
    sold: number
}

export class Bond implements IBond, TDeserializable {
    name: TString;
    amount: BigInteger;
    rate: number;
    used: number;
    sold: number;
    constructor(name: TString, amount: BigInteger, rate: number, used: number, sold: number) {
        this.name = name;
        this.amount = amount;
        this.rate = rate;
        this.used = used;
        this.sold = sold;
    }
    static deserialize(data: Uint8Array): [Bond, Uint8Array] {
        let name;
        [name, data] = TString.deserialize(data);
        const amount = fromLittleEndian(data.slice(0, 8));
        const rate = new BN(data.slice(8, 10), "le").toNumber();
        const used = new BN(data.slice(10, 12), "le").toNumber();
        const sold = new BN(data.slice(12, 14), "le").toNumber();
        return [new Bond(name, amount, rate, used, sold), data.slice(14)];
    }
}
